### PR TITLE
ipc: component: abstract the component creation API

### PR DESCRIPTION
**Pull request by lgirdwood**
Make sure component creation is not tightly coupled to a particular IPC
version. This is mostly a mechanical change of structures being passed
to the comp creation APIs away from IPC specific to general structures.
    
High level changes
      
1) Pass a common component data object and a component
specific data object during create().
    
2) Mark the component IPC derived data as "ipc_config" within the
component device to help developers track the data source origin.
    
3) Pass component specific data during creation so that componets
can allocate and copy to thier private data.
    
4) Comp_dev no longer has component specific data appended to it.
Instead we can store all in the comp private data (and hence use the
compiler to access it rather than by developer access methods).
    
Signed-off-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>

### CODE REVIEW

**Comment by jgunn0262** [comment]
I dont care too much about IPC ABIs, but 2, 3 and 4 seem to make sense to me.

File: src/audio/dai.c [hunk 1]
```diff
@@ -542,7 +549,9 @@ static int dai_params(struct comp_dev *dev,
 	comp_dbg(dev, "dai_params()");
 
 	/* configure dai_data first */
-	dai_data_config(dev);
+	err = dai_data_config(dev);
+	if (!err)
```
Comment (RanderWang) [inline src/audio/dai.c:553]: if (err < 0 ) ?   I validated it on my tgl.

Comment (lgirdwood) [reply src/audio/dai.c]: @RanderWang thanks, that's a time saver for me.

**Review by RanderWang — APPROVED** [review]
LGTM. there are some allocations in each component. Will we plan to allocate it from pipeline memory pool or keep it unchanged ? thanks!

**Comment by lgirdwood** [comment]
> LGTM. there are some allocations in each component. Will we plan to allocate it from pipeline memory pool or keep it unchanged ? thanks!

We can allocate from either resource depending on IPC version and/or topology.

File: src/include/ipc/topology.h [hunk 2]
```diff
@@ -47,6 +47,7 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
+	SOF_COMP_CODEC_ADAPTOR,		/**< codec adaptor */
```
Comment (jgunn0262) [inline src/include/ipc/topology.h:50]: This looks like it also needed in stable branch ?

**Review by jgunn0262 — APPROVED** [review]

**Pull request merged**